# fNUMT (v1.1)
fNUMT is a framework for detecting non-ref NUMTs and the derived NUMT-FPs from short-read whole genome sequencing data

## Getting Started
fNUMT is written in perl

### Prerequisites

#### perl(v5.26.2) and following perl models
* Getopt::Long (https://metacpan.org/pod/Getopt::Long)

- File::Spec (https://metacpan.org/pod/File::Spec)

* File::Basename (https://metacpan.org/pod/File::Basename)

#### Other required softwares
* samtools(1.10)

- cap3(02/10/15)

* blastn(2.13.0+)

All the three softwares are available under the bin folder

***
### Installing
    
    wget https://github.com/sunnyzxh/fNUMT/archive/v1.1.zip
    unzip v1.1.zip
    cd fNUMT-1.1

#### Running the test
    cd bin
    chmod 777 samtools
    chmod 777 cap3
    chmod 777 blastn
    cd ../example
    sh example.sh
    cd example
    
If "DONE SUCCESSFULLY!", there should be example.numts.txt (the non-ref NUMT results), example.nonref_NUMT_mutation.txt (the potential NUMT-FPs) under the folder example

***

### Usage
#### Please prepare the following inputs

* The bam file with reads mapped to chrM (the Revised Cambridge Reference Sequence (rCRS, NC_012920))

- Sample name

* The output direction

Note: 
* The bam file should contain the "SA" tag generated by BWA "mem"

#### Running

perl fNUMT.pl

Function
    
    Detecting non-ref NUMTs and the derived NUMT-FPs from short-read WGS data

Usage
    
    perl fNUMT.pl -b example.bam -s sample.name -o outdir

Options
   
    -h|-help            help
    
    -b|-bam      [s]    bam file
    
    -s|-sample   [s]    sample name
    
    -o|-outdir   [s]    output dir

#### Outputs

* *numts.txt (the non-ref NUMTs result)

- *nonref_NUMT_mutation.txt (the potential NUMT-FPs)

* tmp.reads (the soft-clip reads and discordant read-pairs)

- *seq.for.cap3.fastq.* (sequences, contigs and realignment for each non-ref NUMTs)

***

### Support
Should you have any question, please feel free to contanct sunnyzxh@connect.hku.hk
